sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.ComponentModel;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// Abstraction for a payload of binary data.
+    /// </summary>
+    public struct BinaryData
+    {
+        /// <summary>
+        /// The backing store for the <see cref="BinaryData"/> instance.
+        /// </summary>
+        internal ReadOnlyMemory<byte> Data { get; }
+
+        /// <summary>
+        /// Creates a binary data instance from bytes.
+        /// </summary>
+        /// <param name="data">Byte data.</param>
+        public BinaryData(ReadOnlyMemory<byte> data)
+        {
+            Data = data;
+        }
+
+        /// <summary>
+        /// Creates a binary data instance from a string
+        /// using UTF-8 encoding.
+        /// </summary>
+        /// <param name="data">The string data.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData Create(string data) =>
+            Create(data, Encoding.UTF8);
+
+        /// <summary>
+        /// Creates a binary data instance from a string
+        /// using the specified encoding.
+        /// </summary>
+        /// <param name="data">The string data.</param>
+        /// <param name="encoding">The encoding to use when converting
+        /// the data to bytes.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData Create(string data, Encoding encoding)
+        {
+            Argument.AssertNotNull(encoding, nameof(encoding));
+            return new BinaryData(encoding.GetBytes(data));
+        }
+
+        /// <summary>
+        /// Creates a binary data instance from the specified stream.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData Create(Stream stream) =>
+            CreateAsync(stream, false).EnsureCompleted();
+
+        /// <summary>
+        /// Creates a binary data instance from the specified stream.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static async Task<BinaryData> CreateAsync(Stream stream) =>
```

Comment:
You'll need optional `CancellationToken`s on anything Async.

Reply:
Can add to `CreateAsync`, but the `CreateAsync<T>` and `AsAsync<T>` won't have them as `ObjectSerializer.SerializeAsync/DeserializeAsync` doesn't take a ct.
Will add a TODO.